zrml/rikiddo/src/types/rikiddo_sigmoid_mv.rs

```diff
@@ -0,0 +1,368 @@
+                total_balance = res;
+            } else {
+                return Err("[RikiddoSigmoidMV] Overflow during summation of asset balances");
+            }
+        }
+
+        let denominator = if let Some(res) = fee.checked_mul(total_balance) {
+            res
+        } else {
+            // Highly unlikely and only possible if fee > 100%
+            return Err("[RikiddoSigmoidMV] Overflow during calculation: fee * total_asset_balance");
+        };
+
+        let mut exponents: Vec<FS> = Vec::with_capacity(asset_balances.len());
+        let mut biggest_exponent: FS = FS::from_num(0u8);
+
+        for elem in asset_balances {
+            let exponent = if let Some(res) = elem.checked_div(denominator) {
+                convert_to_signed::<FU, FS>(res)?
+            } else {
+                // Highly unlikely
+                return Err("[RikiddoSigmoidMV] Overflow during calculation: expontent_i = \
+                            asset_balance_i / denominator");
+            };
+
+            if exponent > biggest_exponent {
+                biggest_exponent = exponent;
+            }
+
+            // Panic impossible
+            exponents.push(exponent);
+        }
+
+        // Determine which strategy to use.
+        let biggest_exp_times_log2e = if let Some(res) =
+            self.config.log2_e.checked_mul(biggest_exponent)
+        {
+            res
+        } else {
+            // Highly unlikely
+            return Err("[RikiddoSigmoidMV] Overflow during calculation: log2_e * biggest_exponent");
+        };
+
+        if FS::max_value().int().to_num::<u128>() < asset_balances.len() as u128 {
+            return Err("[RikidoSigmoidMV] Number of assets does not fit in FS");
+        }
+
+        // Panic impossible
+        let log2_number_of_assets: FS =
+            if let Ok(res) = log2::<FS, FS>(FS::from_num(asset_balances.len())) {
+                res
+            } else {
+                // Impossible, since the cost functions checks if elements are present in asset_balances
+                return Err("[RikiddoSigmoidMV] log2(number_of_assets), number_of_assets <= 0");
+            };
+
+        let log2e_times_biggest_exp_plus_log2_num_assets =
+            if let Some(res) = log2_number_of_assets.checked_add(biggest_exp_times_log2e) {
+                res
+            } else {
+                // Highly unlikely
+                return Err("[RikiddoSigmoidMV] Overflow during calculation: biggest_exp * \
+                            log2(e) + log2(num_assets)");
+            };
+
+        let required_bits_minus_one: u128 =
+            if let Some(res) = log2e_times_biggest_exp_plus_log2_num_assets.checked_ceil() {
+                res.to_num()
+            } else {
+                // Highly unlikely
+                return Err("[RikiddoSigmoidMV] Overflow during calculation: ceil(biggest_exp * \
+                            log2(e) + log2(num_assets))");
+            };
```

Comment:
The code contains a couple of checks like this that are either impossible (from my perspective) or so incredibly unlikely, that one has to spend a great bunch of brain-power time to figure out a way to trigger them. Once the code ran for a longer period and we want to keep it like that, those can be removed in a PR that encompasses the optimization of this code.